crypto/passphrase.c

```diff
@@ -42,7 +43,7 @@ int ossl_pw_set_passphrase(struct ossl_passphrase_data_st *data,
     data->type = is_expl_passphrase;
     data->_.expl_passphrase.passphrase_copy =
         OPENSSL_memdup(passphrase, passphrase_len);
-    if (data->_.expl_passphrase.passphrase_copy == NULL) {
+    if (data->_.expl_passphrase.passphrase_copy == NULL && errno == ENOMEM) {
```

Comment:
This feels wrong, because it depends on wether OPENSSL_malloc(0) returns NULL or not,
but that is implementation dependent, and later we have this:
https://github.com/bernd-edlinger/openssl/blob/71396cd048072b69559b46d98cfebfd4474cd712/crypto/passphrase.c#L232-L238
so a NULL will not do the right thing.

Reply:
I think it would be safer to change the statement above to
```C
data->_.expl_passphrase.passphrase_copy =
        passphrase_len != 0 ? OPENSSL_memdup(passphrase, passphrase_len)
                                          : OPENSSL_malloc(1);
```

Reply:
Of course this PR needs a test case.

Reply:
@bernd-edlinger thanks for the suggestion, I incorporated it into the PR and it seems to fix the issue mostly. I have also added some tests but I'm not very sure why I'm still hitting this failure only for the last test where I'm passing a callback:
```js
[Error: error:1400006B:UI routines::processing error] {
  opensslErrorStack: [
    'error:1C80009F:Provider routines::unable to get passphrase',
    'error:07880028:common libcrypto routines::UI lib'
  ],
  library: 'UI routines',
  reason: 'processing error',
  code: 'ERR_OSSL_UI_PROCESSING_ERROR'
}
```
I'm guessing it's an incorrect invocation? PTAL.

Reply:
resolving as the added test passes CI